core/consensus/grandpa/impl/voting_round_impl.cpp

```diff
@@ -335,16 +338,18 @@ namespace kagome::consensus::grandpa {
           break;
         }
 
-        const auto &maybe_finalized = last_round_state.finalized;
+        const auto &maybe_last_finalized = last_round_state.finalized;
+        const auto &maybe_curr_finalized = cur_round_state_.finalized;
+
+        const auto &estimate_is_finalized =
+            [&maybe_estimate](const BlockInfo &finalized) {
+              return maybe_estimate->block_number > finalized.block_number;
+            };
 
-        // we should send primary if last round's state contains finalized
-        // block and last round estimate is better than finalized block
+        // Last round estimate has not been finalized.
         bool should_send_primary =
-            maybe_finalized
-                .map([&maybe_estimate](const BlockInfo &finalized) {
-                  return maybe_estimate->block_number > finalized.block_number;
-                })
-                .value_or(false);
+            maybe_last_finalized.map(estimate_is_finalized).value_or(true)
```

Comment:
Think about to simplify it.

Reply:
How? Without map we would need to write complex if statements. Here everything is inline